File: src/lifecycle_cpp/src/number_publisher.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/int64.hpp"
#include "rclcpp_lifecycle/lifecycle_node.hpp"

using LifeCycleCallBackReturn = rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn;

class NumberPublisherNode : public rclcpp_lifecycle::LifecycleNode
{
public:
    NumberPublisherNode() : LifecycleNode("number_publisher")
    {
        number_ = 1;
        publish_frequency_ = 1.0;
        RCLCPP_INFO(this->get_logger(), "In Constructor");
    }

    LifeCycleCallBackReturn on_configure(const rclcpp_lifecycle::State &previous_state){
        (void)previous_state;
        RCLCPP_INFO(this->get_logger(), "In On Configure");
        number_publisher_ = 
            this->create_publisher<example_interfaces::msg::Int64>("number", 10);
        number_timer_ = 
            this->create_wall_timer(std::chrono::milliseconds((int)(1000.0 / publish_frequency_)),
                                    std::bind(&NumberPublisherNode::publishNumber, this));
        number_timer_->cancel();
        return LifeCycleCallBackReturn::SUCCESS;
    }

    LifeCycleCallBackReturn on_activate(const rclcpp_lifecycle::State &previous_state){
        RCLCPP_INFO(this->get_logger(), "In On Activate");
        number_timer_->reset();
        rclcpp_lifecycle::LifecycleNode::on_activate(previous_state);
        return LifeCycleCallBackReturn::SUCCESS;
    }

    LifeCycleCallBackReturn on_deactivate(const rclcpp_lifecycle::State &previous_state){
        RCLCPP_INFO(this->get_logger(), "In On Deactivate");
        number_timer_->cancel();
        rclcpp_lifecycle::LifecycleNode::on_deactivate(previous_state);
        return LifeCycleCallBackReturn::SUCCESS;
    }

    LifeCycleCallBackReturn on_cleanup(const rclcpp_lifecycle::State &previous_state){
        (void)previous_state;
        RCLCPP_INFO(this->get_logger(), "In On Cleanup");
        number_publisher_.reset();
        number_timer_.reset();
        return LifeCycleCallBackReturn::SUCCESS;
    }
    
    LifeCycleCallBackReturn on_shutdown(const rclcpp_lifecycle::State &previous_state){
        (void)previous_state;
        RCLCPP_INFO(this->get_logger(), "In On Shutdown");
        number_publisher_.reset();
        number_timer_.reset();
        return LifeCycleCallBackReturn::SUCCESS;
    }

    LifeCycleCallBackReturn on_error(const rclcpp_lifecycle::State &previous_state){
        (void)previous_state;
        RCLCPP_INFO(this->get_logger(), "In On Error");
        number_publisher_.reset();
        number_timer_.reset();
        return LifeCycleCallBackReturn::FAILURE;  //Goes to finalized from On Error
    }

private:
    void publishNumber()
    {
        auto msg = example_interfaces::msg::Int64();
        msg.data = number_;
        number_publisher_->publish(msg);
        number_++;
    }

    int number_;
    double publish_frequency_;
    rclcpp::Publisher<example_interfaces::msg::Int64>::SharedPtr number_publisher_;
    rclcpp::TimerBase::SharedPtr number_timer_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NumberPublisherNode>();
    rclcpp::spin(node->get_node_base_interface());
    rclcpp::shutdown();
    return 0;
}

File: src/lifecycle_py/lifecycle_py/number_publisher.py
#!/usr/bin/env python3
import rclpy
from rclpy.lifecycle import LifecycleNode
from rclpy.lifecycle.node import LifecycleState, TransitionCallbackReturn

from example_interfaces.msg import Int64

class NumberPublisherNode(LifecycleNode):
    def __init__(self):
        super().__init__("number_publisher")
        self.get_logger().info("In Constructor")
        self.number_ = 1
        self.publish_frequency_ = 1.0

    # Create ROS2 communications, connect to HW etc.
    def on_configure(self, state: LifecycleState):
        self.get_logger().info("In On Configure")
        self.number_publisher_ = self.create_lifecycle_publisher(Int64, "number", 10)
        self.number_timer_ = self.create_timer(1.0 / self.publish_frequency_, self.publish_number)
        return TransitionCallbackReturn.SUCCESS

    # Destroy ROS2 communications, disconnect HW etc.
    def on_cleanup(self, state: LifecycleState):
        self.get_logger().info("In On CleanUp")
        self.destroy_lifecycle_publisher(self.number_publisher_)
        self.destroy_timer(self.number_timer_)
        return TransitionCallbackReturn.SUCCESS

    def publish_number(self):
        msg = Int64()
        msg.data = self.number_
        self.number_publisher_.publish(msg)
        self.number_ += 1

def main(args=None):
    rclpy.init(args=args)
    node = NumberPublisherNode()
    rclpy.spin(node)
    rclpy.shutdown()
